service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -1,93 +1,208 @@
-import com.epam.esm.dao.TagDAO;
-import com.epam.esm.dto.FilterRequest;
+import com.epam.esm.dao.GiftCertificateDao;
+import com.epam.esm.dao.TagDao;
 import com.epam.esm.entity.GiftCertificate;
 import com.epam.esm.entity.Tag;
-import com.epam.esm.exception.DaoException;
-import com.epam.esm.exceptions.IncorrectParameterException;
+import com.epam.esm.exceptions.*;
 import com.epam.esm.handler.DateHandler;
 import com.epam.esm.service.GiftCertificateService;
 import com.epam.esm.validator.GiftCertificateValidator;
 import com.epam.esm.validator.IdentifiableValidator;
-import org.springframework.beans.factory.annotation.Autowired;
+import com.epam.esm.validator.TagValidator;
+import jakarta.transaction.Transactional;
+import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
+import org.springframework.util.MultiValueMap;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.epam.esm.dao.creator.FilterParameter.*;
+import static com.epam.esm.exceptions.ExceptionMessageKey.GIFT_CERTIFICATE_NOT_FOUND;
 
 @Service
 public class GiftCertificateServiceImpl implements GiftCertificateService {
-    private final GiftCertificateDAO giftCertificateDAO;
-    private final TagDAO tagDAO;
+    private final GiftCertificateDao giftCertificateDao;
+    private final TagDao tagDao;
     private final DateHandler dateHandler;
 
-    @Autowired
-    public GiftCertificateServiceImpl(GiftCertificateDAO giftCertificateDAO, TagDAO tagDAO, DateHandler dateHandler) {
-        this.giftCertificateDAO = giftCertificateDAO;
-        this.tagDAO = tagDAO;
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, TagDao tagDao, DateHandler dateHandler) {
+        this.giftCertificateDao = giftCertificateDao;
+        this.tagDao = tagDao;
         this.dateHandler = dateHandler;
     }
 
     @Override
-    public void insert(GiftCertificate giftCertificate) throws DaoException, IncorrectParameterException {
-        GiftCertificateValidator.validate(giftCertificate);
-        giftCertificateDAO.insert(giftCertificate);
+    @Transactional
+    public GiftCertificate insert(GiftCertificate giftCertificate) {
+        ExceptionResult exceptionResult = new ExceptionResult();
+        GiftCertificateValidator.validate(giftCertificate, exceptionResult);
+        if (!exceptionResult.getExceptionMessages().isEmpty()) {
+            throw new IncorrectParameterException(exceptionResult);
+        }
+        String giftCertificateName = giftCertificate.getName();
+        boolean isGiftCertificateExist = giftCertificateDao.getByName(giftCertificateName).isPresent();
+        if (isGiftCertificateExist) {
+            throw new DuplicateEntityException(ExceptionMessageKey.GIFT_CERTIFICATE_EXIST);
+        }
+        giftCertificate.setCreateDate(dateHandler.getCurrentDate());
```

Comment:
You can use Hibernate envers here for auditing. You don't need to set auditing values manually.